File: implementations/go/harness/no_cross_bc_domain_import.go
package main

import (
	"os"
	"path/filepath"
	"strings"
)

// checkNoCrossBCDomainImport — no-cross-bc-domain-import: from the root
// docs/architecture/tactical-ddd.md, "another Aggregate may only be referenced
// by ID (referencing by object is forbidden)" — this principle applies not
// only between Aggregates within the same Bounded Context
// (no_cross_aggregate_reference.go precisely checks only Payment<->Refund
// inside payment/) but equally between different BCs. domain_layer_isolation.go
// (R1) only checks that domain/ does not import a higher layer
// (application/infrastructure/interface), and no_cross_aggregate_reference.go
// (R5) only looks inside the payment BC — no rule blocks an import between
// different BCs' domain packages, such as `internal/domain/card` directly
// importing `internal/domain/payment`.
//
// It is a violation if any path imported by a file under
// internal/domain/<bc>/*.go is internal/domain/<other-bc> (a BC other than
// itself). A well-formed cross-BC reference is made only through an ID string
// (e.g. `PaymentID string`), so there is no need to import another BC's
// domain package in the first place — the Adapter (infrastructure/acl)
// approach mandated by cross-domain-communication.md is also the
// responsibility of the application/infrastructure layer, not the domain
// layer.
//
// internal/common (shared Value Objects like Money, ID generators) does not
// belong to any specific BC — it is a low-level utility package shared across
// multiple BCs — so it is never matched by the domainImportBC regex (which
// matches only the exact internal/domain/<seg>/ shape) in the first place.
// This repository currently has no "shared type" case that would cause a
// false positive here (every domain/<bc> file imports no domain/ package
// outside its own BC — this rule is a guard against future regressions).
func checkNoCrossBCDomainImport(root string) RuleResult {
	result := RuleResult{Section: "no-cross-bc-domain-import"}
	domainDir := filepath.Join(root, "internal", "domain")
	if _, err := os.Stat(domainDir); os.IsNotExist(err) {
		result.Findings = append(result.Findings, skipFinding("internal/domain/ not found"))
		return result
	}

	found := false
	walkErr := filepath.WalkDir(domainDir, func(path string, d os.DirEntry, err error) error {
		if err != nil || d.IsDir() || !strings.HasSuffix(path, ".go") || strings.HasSuffix(d.Name(), "_test.go") {
			return nil
		}
		slashPath := filepath.ToSlash(path)
		selfMatch := domainImportBC.FindStringSubmatch(slashPath[strings.Index(slashPath, "/internal/domain/"):])
		if selfMatch == nil {
			return nil
		}
		selfBC := selfMatch[1]
		rel, _ := filepath.Rel(root, path)

		imports, parseErr := fileImportPaths(path)
		if parseErr != nil {
			found = true
			result.Findings = append(result.Findings, failFinding(rel, "failed to parse Go file: "+parseErr.Error()))
			return nil
		}

		var otherBCs []string
		for _, imp := range imports {
			m := domainImportBC.FindStringSubmatch("/" + filepath.ToSlash(imp))
			if m != nil && m[1] != selfBC {
				otherBCs = append(otherBCs, m[1])
			}
		}
		found = true
		if len(otherBCs) > 0 {
			result.Findings = append(result.Findings, failFinding(rel,
				"directly imports another Bounded Context's domain package ("+strings.Join(otherBCs, ", ")+") — an Aggregate must reference another Aggregate only by its ID string (docs/architecture/tactical-ddd.md)"))
		} else {
			result.Findings = append(result.Findings, passFinding(rel))
		}
		return nil
	})
	if walkErr != nil {
		result.Findings = append(result.Findings, failFinding(domainDir, "directory walk failed: "+walkErr.Error()))
	} else if !found {
		result.Findings = append(result.Findings, skipFinding("no .go files under internal/domain/<bc>/"))
	}
	return result
}

File: implementations/fastapi/harness/rules/no_cross_bc_domain_import.py
"""[28] no-cross-bc-domain-import: one Bounded Context's domain/ must not directly import
another BC's domain/ (tactical-ddd.md "another Aggregate may only be referenced by ID")

tactical-ddd.md's principle "another Aggregate may only be referenced by ID (an object
reference is forbidden)" applies not only between Aggregates within the same BC (e.g.
payment/Refund must not reference the Payment class directly — checked by the
`no-cross-aggregate-reference` rule), but equally between different BCs.
`domain-layer-isolation` (rules/domain_layer_isolation.py) only checks that domain/
doesn't import the application/·infrastructure/·interface/ layers, and
`no-cross-aggregate-reference` only checks within `payment/domain/{payment.py,refund.py}`
— there was no rule between these two that stopped `src/card/domain/*.py` from directly
importing `src/payment/domain/*` (another BC's domain package). This rule closes that gap.

**Determining the domain**: the segment right before the `domain` segment in the file path
is treated as "the BC this file belongs to" (e.g. `src/card/domain/card.py` → `card`).

**Determining a violation**: it finds the `domain` segment in the import's module path,
and it's a violation if the segment right before it is neither its own BC's name, `src`,
nor a shared-infrastructure directory (`common`/`database`/`outbox`/`task_queue`/`config`,
`common.is_shared_dir`). A relative import within the same BC
(`from .account_status import ...`, `from .errors import ...`) is never a false positive,
since the module string has no `domain` segment at all (a Python relative import never
repeats the package name within the same package). Referencing a shared util like `from
...common.generate_id import generate_id` is also never a false positive, since `common`
is treated as a shared directory.
"""

from __future__ import annotations

import ast

from .common import RuleResult, failed, is_shared_dir, norm, passed, read, rel, skipped


def _own_bc(fn: str) -> str | None:
    parts = fn.split("/")
    if "domain" not in parts:
        return None
    idx = parts.index("domain")
    if idx == 0:
        return None
    return parts[idx - 1]


def _module_names(node: ast.AST) -> list[str]:
    if isinstance(node, ast.ImportFrom):
        return [node.module] if node.module else []
    if isinstance(node, ast.Import):
        return [alias.name for alias in node.names]
    return []


def check(root: str, py_files: list[str]) -> RuleResult:
    result = RuleResult("no-cross-bc-domain-import")
    found = False
    for f in py_files:
        fn = norm(f)
        if "/domain/" not in fn:
            continue
        own_bc = _own_bc(fn)
        if own_bc is None:
            continue
        found = True
        r = rel(root, f)
        src = read(f)

        try:
            tree = ast.parse(src, filename=f)
        except SyntaxError as e:
            result.add(failed(r, f"Failed to parse the file: {e}"))
            continue

        violations: list[str] = []
        for node in ast.walk(tree):
            if not isinstance(node, (ast.ImportFrom, ast.Import)):
                continue
            for module in _module_names(node):
                segments = module.split(".")
                if "domain" not in segments:
                    continue
                idx = segments.index("domain")
                prefix = segments[:idx]
                if not prefix:
                    continue  # a same-BC relative import ("domain.x") — no BC name
                other_bc = prefix[-1]
                if other_bc in {"src", own_bc} or is_shared_dir(other_bc):
                    continue
                violations.append(f"{module} ({other_bc} BC)")

        if violations:
            result.add(
                failed(
                    r,
                    f"domain/({own_bc}) directly imports another BC's domain/ — another Aggregate"
                    " may only be referenced by ID; an object reference is forbidden (tactical-ddd.md): "
                    + "; ".join(violations),
                )
            )
        else:
            result.add(passed(f"{r} (no-cross-bc-domain-import)"))
    if not found:
        result.add(skipped("No Python file in domain/"))
    return result
